### ISSUE TITLE
Check Fork failed with error code 11 (Resource temporarily unavailable)

### ISSUE THREAD

**Issue by akqopensystems** [Icinga/icinga2#6056]
Cannot fork check processes

## Current Behavior

Icinga2 Satellite can't fork new check processes and fails with error message:
```[2018-02-01 16:19:22 +0100] critical/Process: Fork failed with error code 11 (Resource temporarily unavailable)
Context:
        (0) Executing check for object '<hostname>!ippbxsrv'

[2018-02-01 16:19:22 +0100] warning/PluginCheckTask: Check command for object '<hostname>!ippbxsrv' (PID: -1, arguments: '/usr/lib64/nagios/plugins/check_nrpe' '-H' '<ip>' '-c' 'check_service_<hostname>' '-t' '60' '-a' 'IpPbxSrv') terminated with exit code 128, output: Fork failed with error code 11 (Resource temporarily unavailable)
```
Version is:
```
[root@xxxxmo06 icinga2.service.d]# icinga2 --version
icinga2 - The Icinga 2 network monitoring daemon (version: r2.8.1-1)

Copyright (c) 2012-2017 Icinga Development Team (https://www.icinga.com/)
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl2.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Application information:
  Installation root: /usr
  Sysconf directory: /etc
  Run directory: /run
  Local state directory: /var
  Package data directory: /usr/share/icinga2
  State path: /var/lib/icinga2/icinga2.state
  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
  Objects path: /var/cache/icinga2/icinga2.debug
  Vars path: /var/cache/icinga2/icinga2.vars
  PID path: /run/icinga2/icinga2.pid

System information:
  Platform: Red Hat Enterprise Linux Server
  Platform version: 7.4 (Maipo)
  Kernel: Linux
  Kernel version: 3.10.0-693.11.6.el7.x86_64
  Architecture: x86_64

Build information:
  Compiler: GNU 4.8.5
  Build host: unknown
```
Same behaviour with Icinga2 Version 2.8.0.
Systemd-Service is configured with 
```
TasksMax=infinity
LimitNPROC=62883
LimitNOFILE=50000
```
At the moment, we are monitoring ~5000 hosts with ~18000 services.

After a restart of the icinga2 service, all works well for about 2 hours, then the errors start to appear again.

## Your Environment
[root@xxxxxmo06 icinga2.service.d]# icinga2 feature list
critical/Application: Error: Function call 'fork' failed with error code 11, 'Resource temporarily unavailable'

Additional information is available in '/var/log/icinga2/crash/report.1517498852.168836'

Aborted
[root@xxxxmo06 icinga2.service.d]# cat /var/log/icinga2/crash/report.1517498852.168836
Caught unhandled exception.
Current time: 2018-02-01 16:27:32 +0100

Application information:
  Application version: r2.8.1-1
  Installation root: /usr
  Sysconf directory: /etc
  Run directory: /run
  Local state directory: /var
  Package data directory: /usr/share/icinga2
  State path: /var/lib/icinga2/icinga2.state
  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
  Objects path: /var/cache/icinga2/icinga2.debug
  Vars path: /var/cache/icinga2/icinga2.vars
  PID path: /run/icinga2/icinga2.pid

System information:
  Platform: Red Hat Enterprise Linux Server
  Platform version: 7.4 (Maipo)
  Kernel: Unknown
  Kernel version: Unknown
  Architecture: Unknown

Build information:
  Compiler: GNU 4.8.5
  Build host: unknown

Error: Function call 'fork' failed with error code 11, 'Resource temporarily unavailable'

***
* This would indicate a runtime problem or configuration error. If you believe this is a bug in Icinga 2
* please submit a bug report at https://github.com/Icinga/icinga2 and include this stack trace as well as any other
* information that might be useful in order to reproduce this problem.
***
:(

[root@xxxxxmo06 crash]# systemctl restart icinga2
[root@xxxxxmo06 crash]# icinga2 feature list
Disabled features: command compatlog debuglog elasticsearch gelf graphite influxdb livestatus notification opentsdb perfdata statusdata syslog
Enabled features: api checker mainlog

Any ideas where to look?

**Comment by dgoetz** [comment]
My best guess would be number of open files, but 50000 should be enough. You can find the number of open files by looking at /proc/PID/fd like `sudo ls -l /proc/1/fd/ | wc -l`, the PID would be the one of the second icinga2 process I think.

**Comment by Crunsher** [comment]
I think @dgoetz is right, this sounds like a problem with too many open files. You can also run `sysctl fs.file-nr` (thel 7 runs systemd, right?) to see the global limit, in case some other task is hogging all the file descriptors.

**Comment by akqopensystems** [comment]
Hello,

the problem is temporarily solved. It is a problem of selinux, which could not execute a sigkill.
Here a excerpt from the log. At the moment we set selinux in permissive mode.

In log we see about 140000 selinux violations. 

**-- snip --**
Jan 23 17:18:03 xxxxxmo06 python: SELinux is preventing /usr/lib64/icinga2/sbin/icinga2 from using the sigkill access on a process.#012#012*****  Plugin catchall (100. confidence) suggests   **************************#012#012If you believe that icinga2 should be allowed sigkill access on processes labeled nagios_unconfined_plugin_t by default.#012Then you should report this as a bug.#012You can generate a local policy module to allow this access.#012Do#012allow this access for now by executing:#012# ausearch -c 'icinga2' --raw | audit2allow -M my-icinga2#012# semodule -i my-icinga2.pp#012


Additional Information:
Source Context                system_u:system_r:icinga2_t:s0
Target Context                system_u:system_r:nagios_unconfined_plugin_t:s0
Target Objects                Unknown [ process ]
Source                        icinga2
Source Path                   /usr/lib64/icinga2/sbin/icinga2
Port                          <Unknown>
Host                          xxxxxmo06
Source RPM Packages           icinga2-bin-2.8.1-1.el7.icinga.x86_64
Target RPM Packages           
Policy RPM                    selinux-policy-3.13.1-166.el7_4.7.noarch
Selinux Enabled               True
Policy Type                   targeted
Enforcing Mode                Enforcing
Host Name                     xxxxxmo06
Platform                      Linux xxxxxmo06
                              3.10.0-693.11.6.el7.x86_64 #1 SMP Thu Dec 28
                              14:23:39 EST 2017 x86_64 x86_64
Alert Count                   62640
First Seen                    2018-01-06 04:30:35 CET
Last Seen                     2018-02-02 08:44:05 CET
Local ID                      745345ab-8aa4-4137-8183-ead208b8413f

Raw Audit Messages
type=AVC msg=audit(1517557445.718:164): avc:  denied  { sigkill } for  pid=2066 comm="icinga2" scontext=system_u:system_r:icinga2_t:s0 tcontext=system_u:system_r:nagios_unconfined_plugin_t:s0 tclass=process


type=AVC msg=audit(1517557445.718:164): avc:  denied  { sigkill } for  pid=2066 comm="icinga2" scontext=system_u:system_r:icinga2_t:s0 tcontext=system_u:system_r:nagios_unconfined_plugin_t:s0 tclass=process


type=SYSCALL msg=audit(1517557445.718:164): arch=x86_64 syscall=kill success=no exit=EACCES a0=ffffdfae a1=9 a2=ffffffff a3=fffffe00 items=0 ppid=1 pid=2066 auid=4294967295 uid=995 gid=992 euid=995 suid=995 fsuid=995 egid=992 sgid=992 fsgid=992 tty=(none) ses=4294967295 comm=icinga2 exe=/usr/lib64/icinga2/sbin/icinga2 subj=system_u:system_r:icinga2_t:s0 key=(null)

Hash: icinga2,icinga2_t,nagios_unconfined_plugin_t,process,sigkill
**-- snip --**

**Comment by dgoetz** [comment]
So it is a duplicate of #5916, which I already solved and is planned for 2.8.2. 

If you want not the complete system in permissive mode you can set only the icinga2_t domain to permissive or build the new policy on your own for now.
